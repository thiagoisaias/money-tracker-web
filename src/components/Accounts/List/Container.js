import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";

import { accountType } from "types";

import List from "./List";

import {
  fetchAccounts,
  clearAccountsError
} from "store/actions/accounts/accounts";

export class Container extends Component {
  static propTypes = {
    accountList: PropTypes.arrayOf(accountType).isRequired,
    error: PropTypes.string,
    isLoading: PropTypes.bool.isRequired,
    fetchAccounts: PropTypes.func.isRequired
  };

  componentDidMount() {
    const { fetchAccounts } = this.props;

    fetchAccounts();
  }

  render() {
    const { accountList, error, isLoading } = this.props;
    const propsToPass = {
      accountList,
      error,
      isLoading
    };
    return <List {...propsToPass} />;
  }

  componentWillUnmount() {
    const { error, clearAccountsError } = this.props;

    if (error) {
      clearAccountsError();
    }
  }
}

const mapStateToProps = state => ({
  isLoading: state.accounts.isLoading,
  error: state.accounts.error,
  accountList: state.accounts.accountList
});

const mapDispatchToProps = {
  fetchAccounts,
  clearAccountsError
};

export default connect(mapStateToProps, mapDispatchToProps)(Container);
