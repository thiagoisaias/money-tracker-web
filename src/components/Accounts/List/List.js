import React, { Fragment } from "react";
import PropTypes from "prop-types";

import {
  Wrapper,
  Header,
  Title,
  LinkButton,
  ErrorMessage,
  BlankStateMessage
} from "./styled";

import ItemContainer from "../Item/ItemContainer";
import Layout from "components/Layout/Layout";
import Spinner from "components/UI/Spinner/Spinner";
import withExpandableItem from "hoc/withExpandableItem/withExpandableItem";

const List = props => {
  const {
    accountList,
    activeItemId,
    handleActiveItem,
    error,
    isLoading
  } = props;

  const list = accountList.map(account => {
    const isActive = activeItemId === account.id;
    return (
      <ItemContainer
        key={account.id}
        handleActiveItem={handleActiveItem}
        isActive={isActive}
        accountData={account}
      />
    );
  });

  return (
    <Layout>
      <Wrapper>
        <Fragment>
          <Header>
            <Title> {"Accounts"} </Title>
            <LinkButton to="/accounts/new">{"Add Account"}</LinkButton>
          </Header>
          {error && <ErrorMessage>{error}</ErrorMessage>}
          {list.length === 0 &&
            !error &&
            !isLoading && (
              <BlankStateMessage>
                &#9888;{"There are no accounts registered yet."}
              </BlankStateMessage>
            )}
          {isLoading ? <Spinner size={25} color={"#777"} /> : list}
        </Fragment>
      </Wrapper>
    </Layout>
  );
};

List.propTypes = {
  accountList: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      initialBalance: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired
    })
  ).isRequired,
  activeItemId: PropTypes.number,
  error: PropTypes.string,
  isLoading: PropTypes.bool.isRequired,
  handleActiveItem: PropTypes.func.isRequired
};

export default withExpandableItem(List);
