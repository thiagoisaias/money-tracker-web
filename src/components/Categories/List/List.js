import React, { Fragment } from "react";
import PropTypes from "prop-types";

import { categoryType } from "types";

import {
  Wrapper,
  Header,
  Title,
  LinkButton,
  ErrorMessage,
  BlankStateMessage
} from "./styled";

import ItemContainer from "../Item/Container";
import Layout from "components/Layout/Layout";
import Spinner from "shared/Spinner/Spinner";
import withExpandableItem from "hoc/withExpandableItem/withExpandableItem";

const List = props => {
  const {
    activeItemId,
    categoryList,
    error,
    handleActiveItem,
    isLoading
  } = props;

  const list = categoryList.map(category => {
    const isActive = activeItemId === category.id;
    return (
      <ItemContainer
        key={category.id}
        handleActiveItem={handleActiveItem}
        isActive={isActive}
        categoryData={category}
      />
    );
  });

  return (
    <Layout>
      <Wrapper>
        <Fragment>
          <Header>
            <Title> {"Categories"} </Title>
            <LinkButton to="/categories/new">{"Add Category"}</LinkButton>
          </Header>
          {error && <ErrorMessage>{error}</ErrorMessage>}
          {list.length === 0 &&
            !error &&
            !isLoading && (
              <BlankStateMessage>
                &#9888;{"There are no accounts registered yet."}
              </BlankStateMessage>
            )}
          {isLoading ? <Spinner size={25} color={"#777"} /> : list}
        </Fragment>
      </Wrapper>
    </Layout>
  );
};

List.propTypes = {
  categoryList: PropTypes.arrayOf(categoryType).isRequired,
  activeItemId: PropTypes.number,
  error: PropTypes.string,
  isLoading: PropTypes.bool.isRequired,
  handleActiveItem: PropTypes.func.isRequired
};

export default withExpandableItem(List);
