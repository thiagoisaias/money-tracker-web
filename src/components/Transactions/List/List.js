import React from "react";
import PropTypes from "prop-types";

import {
  Wrapper,
  Header,
  LinkButton,
  MonthSelector,
  Arrow,
  CurrentMonth,
  BlankStateMessage,
  ErrorMessage
} from "./styled";

import ItemContainer from "../Item/Container";
import Spinner from "shared/Spinner/Spinner";

const List = props => {
  const {
    activeItemId,
    error,
    isLoading,
    handleActiveItem,
    handleMonthChange,
    selectedDate,
    transactionList
  } = props;

  const transactions = transactionList.map(transaction => {
    const isActive = activeItemId === transaction.id;
    return (
      <ItemContainer
        key={transaction.id}
        handleActiveItem={handleActiveItem}
        isActive={isActive}
        transactionData={transaction}
      />
    );
  });

  return (
    <Wrapper>
      <Header>
        <LinkButton to="/transactions/new">{"Add Transaction"}</LinkButton>
        <MonthSelector>
          <Arrow
            onClick={() => {
              handleMonthChange("previous");
            }}
          >
            &lsaquo;
          </Arrow>
          <CurrentMonth>{selectedDate.format("MMMM of YYYY")}</CurrentMonth>
          <Arrow
            onClick={() => {
              handleMonthChange("next");
            }}
          >
            &rsaquo;
          </Arrow>
        </MonthSelector>
      </Header>
      {error && <ErrorMessage>{error}</ErrorMessage>}
      {transactions.length === 0 &&
        !error &&
        !isLoading && (
          <BlankStateMessage>
            &#9888;{"There are no transactions registered yet."}
          </BlankStateMessage>
        )}
      {isLoading ? <Spinner size={25} color={"#777"} /> : transactions}
    </Wrapper>
  );
};

List.propTypes = {
  activeItemId: PropTypes.number,
  error: PropTypes.string,
  isLoading: PropTypes.bool.isRequired,
  handleActiveItem: PropTypes.func.isRequired,
  handleMonthChange: PropTypes.func.isRequired,
  selectedDate: PropTypes.object.isRequired,
  transactionList: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      account: PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        initialBalance: PropTypes.number.isRequired
      }).isRequired,
      category: PropTypes.shape({
        id: PropTypes.number.isRequired,
        color: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired
      }).isRequired,
      date: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      transactionType: PropTypes.string.isRequired,
      value: PropTypes.number.isRequired
    })
  )
};

export default List;
