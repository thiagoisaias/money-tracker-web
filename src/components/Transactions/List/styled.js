import styled from "styled-components";
import { devices } from "utils/devices";
import { NavLink } from "react-router-dom";

export const Wrapper = styled.div`
  background-color: #fff;
  box-shadow: 1px;
  margin: 0 auto;
  color: #484848;
  box-shadow: 0.5px 1px 1px 1px #ddd;

  @media ${devices.small} {
    padding: 16px;
    margin-top: 32px;
  }

  @media ${devices.mediumUp} {
    padding: 32px;
    margin: 32px auto;
    border-radius: 2px;
  }
`;

export const Header = styled.div`
  display: flex;
  position: relative;
  align-items: center;
  margin-bottom: 32px;
  padding-bottom: 16px;
  border-bottom: 1px solid #f2f2f2;

  @media ${devices.small} {
    flex-direction: column;
  }
`;

export const LinkButton = styled(NavLink)`
  display: block;
  width: 130px;
  height: 40px;
  line-height: 40px;
  border-radius: 2px;
  background-color: #333;
  color: #fff;
  font-size: 12px;
  text-align: center;

  &:hover {
    cursor: pointer;
    opacity: 0.95;
  }

  @media ${devices.small} {
    margin-top: 8px;
    margin-bottom: 24px;
  }
`;

export const MonthSelector = styled.div`
  display: flex;
  justify-content: center;
  font-size: 16px;
  line-height: 32px;
  font-weight: 600;
  letter-spacing: 0.5px;
  cursor: default;

  @media ${devices.mediumUp} {
    position: absolute;
    left: calc(50% - 153.28px);
  }
`;

export const Arrow = styled.div`
  margin: 0 24px;
  text-align: center;
  width: 32px;
  font-size: 32px;
  line-height: 27px;
  font-weight: 600;
  user-select: none;
  &:hover {
    cursor: pointer;
    opacity: 0.7;
  }
`;

export const CurrentMonth = styled.div`
  width: 160px;
  text-align: center;
`;

export const BlankStateMessage = styled.p`
  display: block;
  color: #999;
  font-size: 14px;
  font-weight: 600;
  text-align: center;
  margin-top: 55px;
  margin-bottom: 16px;
`;
