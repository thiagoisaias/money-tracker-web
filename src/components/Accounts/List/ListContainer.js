import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";

import List from "./List";

import { fetchAccounts } from "store/actions/accounts/accounts";

export class ListContainer extends Component {
  componentDidMount() {
    const { onFetchAccountList } = this.props;

    onFetchAccountList();
  }

  render() {
    const { accountList, error, isLoading } = this.props;
    const propsToPass = {
      accountList,
      error,
      isLoading
    };
    return <List {...propsToPass} />;
  }
}

ListContainer.propTypes = {
  accountList: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      initialBalance: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired
    })
  ).isRequired,
  error: PropTypes.string,
  isLoading: PropTypes.bool.isRequired,
  onFetchAccountList: PropTypes.func.isRequired
};

const mapStateToProps = state => {
  return {
    isLoading: state.accounts.isLoading,
    error: state.accounts.error,
    accountList: state.accounts.accountList
  };
};

const mapDispatchToProps = dispatch => {
  return {
    onFetchAccountList: () => {
      dispatch(fetchAccounts());
    }
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ListContainer);
