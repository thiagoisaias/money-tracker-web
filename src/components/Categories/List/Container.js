import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";

import {
  clearCategoriesError,
  fetchCategories
} from "store/actions/categories/categories";

import List from "./List";

export class Container extends Component {
  render() {
    const { categoryList, error, isLoading } = this.props;
    return (
      <List categoryList={categoryList} error={error} isLoading={isLoading} />
    );
  }

  componentDidMount() {
    const { onFetchCategoryList } = this.props;
    onFetchCategoryList();
  }

  componentWillUnmount() {
    const { error, onClearCategoriesError } = this.props;

    if (error) {
      onClearCategoriesError();
    }
  }
}

Container.propTypes = {
  categoryList: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      color: PropTypes.string.isRequired
    })
  ).isRequired,
  error: PropTypes.string,
  isLoading: PropTypes.bool.isRequired,
  onFetchCategoryList: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  categoryList: state.categories.categoryList,
  error: state.categories.error,
  isLoading: state.categories.isLoading
});

const mapDispatchToProps = {
  onClearCategoriesError: clearCategoriesError,
  onFetchCategoryList: fetchCategories
};

export default connect(mapStateToProps, mapDispatchToProps)(Container);
