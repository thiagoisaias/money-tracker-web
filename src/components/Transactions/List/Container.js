import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import moment from "moment";

import List from "./List";
import withExpandableItem from "hoc/withExpandableItem/withExpandableItem";

import { fetchTransactions } from "store/actions/transactions/transactions";

export class Container extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selectedDate: this.props.selectedDate || moment()
    };
  }

  handleMonthChange = direction => {
    switch (direction) {
      case "next":
        this.handleNextMonth();
        break;
      case "previous":
        this.handlePreviousMonth();
        break;
      default:
        return;
    }
  };

  handleNextMonth = () => {
    this.setState(prevState => {
      return { selectedDate: prevState.selectedDate.clone().add(1, "month") };
    });
  };

  handlePreviousMonth = () => {
    this.setState(prevState => {
      return {
        selectedDate: prevState.selectedDate.clone().subtract(1, "month")
      };
    });
  };

  componentDidMount() {
    const { onFetchTransactionList } = this.props;
    onFetchTransactionList();
  }

  render() {
    const {
      activeItemId,
      error,
      isLoading,
      handleActiveItem,
      transactionList
    } = this.props;
    return (
      <List
        activeItemId={activeItemId}
        error={error}
        isLoading={isLoading}
        handleActiveItem={handleActiveItem}
        handleMonthChange={this.handleMonthChange}
        selectedDate={this.state.selectedDate}
        transactionList={transactionList}
      />
    );
  }
}

Container.propTypes = {
  activeItemId: PropTypes.number,
  error: PropTypes.string,
  isLoading: PropTypes.bool.isRequired,
  handleActiveItem: PropTypes.func.isRequired,
  onFetchTransactionList: PropTypes.func.isRequired,
  selectedDate: PropTypes.object,
  transactionList: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      account: PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        initialBalance: PropTypes.number.isRequired
      }).isRequired,
      category: PropTypes.shape({
        id: PropTypes.number.isRequired,
        color: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired
      }).isRequired,
      date: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      transactionType: PropTypes.string.isRequired,
      value: PropTypes.number.isRequired
    })
  )
};

const mapStateToProps = state => {
  return {
    error: state.transactions.error,
    isLoading: state.transactions.isLoading,
    transactionList: state.transactions.transactionList
  };
};

const mapDispatchToProps = dispatch => {
  return {
    onFetchTransactionList: () => {
      dispatch(fetchTransactions());
    }
  };
};

export default withExpandableItem(
  connect(mapStateToProps, mapDispatchToProps)(Container)
);
