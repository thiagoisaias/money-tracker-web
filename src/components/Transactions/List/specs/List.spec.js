import React from "react";
import ReactDOM from "react-dom";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import moment from "moment";

import List from "../List";

describe("Transaction List", () => {
  const mockProps = {
    activeItemId: 1,
    error: null,
    isLoading: false,
    handleActiveItem: jest.fn(),
    handleMonthChange: jest.fn(),
    selectedDate: moment("2018-05-29", "YY-MM-DD"),
    transactionList: [
      {
        id: 1,
        account: {
          id: 1,
          name: "Lorem",
          initialBalance: 0
        },
        date: "2018-04-22",
        description: "Uber",
        value: 150.67,
        category: {
          id: 1,
          name: "Transport",
          color: "#DDD"
        },
        transactionType: "expense"
      },
      {
        id: 2,
        account: {
          id: 2,
          name: "Lorem",
          initialBalance: 0
        },
        date: "2018-04-23",
        description: "Lorem Ipsum Dolor Transaction Mousepad Monitor",
        value: 200.43,
        category: {
          id: 2,
          name: "Salary",
          color: "lightblue"
        },
        transactionType: "earning"
      }
    ]
  };
  const component = shallow(<List {...mockProps} />);

  it("renders properly", () => {
    expect(toJson(component)).toMatchSnapshot();
  });
});
