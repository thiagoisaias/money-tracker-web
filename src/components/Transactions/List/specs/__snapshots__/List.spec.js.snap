// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transaction List renders properly 1`] = `
<styled.div>
  <styled.div>
    <Styled(NavLink)
      to="/transactions/new"
    >
      Add Transaction
    </Styled(NavLink)>
    <styled.div>
      <styled.div
        onClick={[Function]}
      >
        ‹
      </styled.div>
      <styled.div>
        May of 2020
      </styled.div>
      <styled.div
        onClick={[Function]}
      >
        ›
      </styled.div>
    </styled.div>
  </styled.div>
  <withRouter(Connect(Container))
    handleActiveItem={[Function]}
    isActive={true}
    key="1"
    transactionData={
      Object {
        "account": Object {
          "id": 1,
          "initialBalance": 0,
          "name": "Lorem",
        },
        "category": Object {
          "color": "#DDD",
          "id": 1,
          "name": "Transport",
        },
        "date": "2018-04-22",
        "description": "Uber",
        "id": 1,
        "transactionType": "expense",
        "value": 150.67,
      }
    }
  />
  <withRouter(Connect(Container))
    handleActiveItem={[Function]}
    isActive={false}
    key="2"
    transactionData={
      Object {
        "account": Object {
          "id": 2,
          "initialBalance": 0,
          "name": "Lorem",
        },
        "category": Object {
          "color": "lightblue",
          "id": 2,
          "name": "Salary",
        },
        "date": "2018-04-23",
        "description": "Lorem Ipsum Dolor Transaction Mousepad Monitor",
        "id": 2,
        "transactionType": "earning",
        "value": 200.43,
      }
    }
  />
</styled.div>
`;
